From 33e61b5d04488804e3ec6825e1b6587bdd6f549b Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:55 -0400
Subject: [fs] nfsd: Make lock stateid take a reference to the lockowner

Message-id: <1410898849-22126-112-git-send-email-bfields@redhat.com>
Patchwork-id: 95727
O-Subject: [Patch RHEL7 111/165] nfsd: Make lock stateid take a reference to the lockowner
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

A necessary step toward client_mutex removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: e4f1dd7fc23f92efdaaa07d001b3dd5391505580
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 906c860..88225f0 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -928,6 +928,8 @@ static void nfs4_free_lock_stateid(struct nfs4_stid *stid)
  file = find_any_file(stp->st_stid.sc_file);
  if (file)
   filp_close(file, (fl_owner_t)lo);
+ if (stp->st_stateowner)
+  nfs4_put_stateowner(stp->st_stateowner);
  nfs4_free_ol_stateid(stid);
 }
 
@@ -4831,6 +4833,7 @@ init_lock_stateid(struct nfs4_ol_stateid *stp, struct nfs4_lockowner *lo,
  atomic_inc(&stp->st_stid.sc_count);
  stp->st_stid.sc_type = NFS4_LOCK_STID;
  stp->st_stateowner = &lo->lo_owner;
+ atomic_inc(&lo->lo_owner.so_count);
  get_nfs4_file(fp);
  stp->st_stid.sc_file = fp;
  stp->st_stid.sc_free = nfs4_free_lock_stateid;
-- 
1.7.1